### ISSUE TITLE
Fetch API overhead

### ISSUE THREAD

**Issue by bjoe2k4** [jhabdas/fetch-inject#17]
At least on current versions of Firefox and Chrome, a second connection is established when using the Fetch API, adding additional overhead (+SSL handshake) to `fetch-inject`. Suprisingly, the second connection is also established when using `h2`.Therefore, on a small sites with small number of js/css assets to be fetched, fetch-inject might be actually slower than traditional non-blocking methods to load js/css on the the first page load. Only after these two connections are established (e.g. page reload), `fetch-inject` performs considerably better.

A corresponding note in the readme would be appreciated.

**Comment by jhabdas** [comment]
@bjoe2k4 Thanks for the info. There are many edge and corner cases for one to consider when adopting new technology, and I'd rather drive individuals to make their own decisions using a testing methodology than try and keep pace with changing browser behavior. Could you provide a reduced test case showing the behavior and consider it in context of HTTP/2 with server push to help make your desire and observations more clear to those searching through issues?

**Comment by bjoe2k4** [comment]
Sorry, I don't have access to a server-push enabled webserver, since nginx doesn't support push yet.
Tbh, this isn't really an edge case, I cannot really imagine a more common use case other than the first page load without previously established connections.

fetch-inject:
![grafik](https://cloud.githubusercontent.com/assets/1668390/25579136/3ec0fb50-2e75-11e7-8556-c0d3f91944f0.png)

plain inject via dynamically injected script.src element:
![grafik](https://cloud.githubusercontent.com/assets/1668390/25579146/64180ace-2e75-11e7-9912-c7115c7257bf.png)

According to this, I observe a ~40% slower page load using fetch-inject.

**Comment by jhabdas** [comment]
Reopening for now for visibility. That's arguably better than README. Thanks for spotting. Next question. Would you use Fetch Injection to load a site with only one external asset?

**Comment by bjoe2k4** [comment]
Well, there is no reason keeping users from doing so according to the current readme. Actually, with more assets fetch-inject comes closer, but still all the benefit is eaten up by the additional connection.

**First connect**
fetch-inject:
![grafik](https://cloud.githubusercontent.com/assets/1668390/25580533/01cf942e-2e82-11e7-9b9f-e7d3ae574bbe.png)
inject:
![grafik](https://cloud.githubusercontent.com/assets/1668390/25580544/12087478-2e82-11e7-846c-e48b97fcd88d.png)

**With connections already established**
fetch-inject:
![grafik](https://cloud.githubusercontent.com/assets/1668390/25580308/49a3fd46-2e80-11e7-9ff5-27588396733e.png)
inject:
![grafik](https://cloud.githubusercontent.com/assets/1668390/25580330/7c13b960-2e80-11e7-90b2-c3329ed1918e.png)

**Comment by jhabdas** [comment]
@bjoe2k4 could you share some code snippets to go along with those graphs?

**Comment by jhabdas** [comment]
Strike that. Referencing above request:

> Could you provide a reduced test case showing the behavior [....]

Graphs are great, but to have an effective conversation here we need to be doing something useful like loading jQuery and several dependences as stated in the use cases.

**Comment by bjoe2k4** [comment]
I don't see what the difference it would make when replacing jquery in the above example with `$younameit`. Even with dependencies applied (order of script execution), I cannot measure a significant difference.

**Comment by jhabdas** [comment]
> Even with dependencies applied (order of script execution), I cannot measure a significant difference.

Please feel free to dismiss use of Fetch Injection if it does not suit your needs. If you have questions, comments or concerns about the Fetch API itself you can find the living standard in code here:

In summary, you are noticing some overhead using Fetch API, did not provide a reduced test case, showed no code snippets and have not pointed to anything constructive help advance this project with this issue.

Here are some things you may find useful as you move forward with or without Fetch Injecton:

- https://fetch.spec.whatwg.org/
- https://github.com/whatwg/fetch
- https://github.com/kangax/compat-table

Cheers and thanks helping kick the tires.

**Comment by bjoe2k4** [comment]
> Strike that.

I have probably misinterpreted this previous comment. Of course I am happy to provide my test cases. Code of my test cases are as follows.

fetch-inject.html
```
<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><title>title</title>
<script>
var fetchInject = *insert yourself*

fetchInject([
'/js/jquery.min.js?v=9'],fetchInject([
'/js/jquery.min.js?v=8'],fetchInject([
'/js/jquery.min.js?v=7'],fetchInject([
'/js/jquery.min.js?v=6'],fetchInject([
'/js/jquery.min.js?v=5'],fetchInject([
'/js/jquery.min.js?v=4'],fetchInject([
'/js/jquery.min.js?v=3'],fetchInject([
'/js/jquery.min.js?v=2'],fetchInject([
'/js/jquery.min.js?v=1'])))))))));

</script>
</head><body>test</body></html>
```
inject.html:
```
<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><title>title</title>
<script>
[1,2,3,4,5,6,7,8,9].forEach(function(el) {
    var g = document.createElement('script'); g.type = 'text/javascript'; g.async = false; g.src = '/js/jquery.min.js?v='+el;document.head.appendChild(g);
});
</script>
</head><body>test</body></html>
```

I opened this issue because I am out of ideas how to improve on this since the spec is kinda nailed down and implemented in browsers already.

**Comment by jhabdas** [comment]
> I opened this issue because I am out of ideas how to improve on this since the spec is kinda nailed down and implemented in browsers already.

Try taking a stab at this. It has much more value to the Web as a whole than this library:
https://github.com/kangax/compat-table/issues/1052

**Comment by bjoe2k4** [comment]
Just a quick update on this issue: according to [this](https://fetch.spec.whatwg.org/#connections) the connection is demtermined based on origin and credentials flag (bool). Credentials here include cookies and authorization http headers.
Since fetch request credentials default to _false_, a new connection is made even when the asset is located on the same server and an existing connection could have been reused. A possible way to omit the additional connection would be to set credentials to true like in the following:
```
fetch('/whatever', {
  credentials: 'include'
})
```
However, external requests would be blocked by CORS this way. So we need to know beforehand if the request is gonna be external or not and set the credentials header accordingly.

@jhabdas Would you be interested in such an improvement?

**Comment by jhabdas** [comment]
@bkjoe2k4 Thank you kindly for the research. I will reopen this issue now to work against in PR. Please focus on 80/20, and keep the code tight and extensible. If you run into any spec issues we'll open issues as needed so we keep implementation lean and mean. And I deeply appreciate your continued interest in Fetch Inject.

**Comment by jhabdas** [comment]
Blocked by #4

**Comment by jhabdas** [comment]
Closing due to inactivity.

**Comment by jhabdas** [comment]
The following waterfalls have been added to the README with a link to a live demo running on HTTP/2.

![fetch-inject-serviceworker-caching](https://cloud.githubusercontent.com/assets/440298/26752598/2331989c-4886-11e7-9ae6-66ff66e92099.png)
![fetch-inject-unprimed-cache](https://cloud.githubusercontent.com/assets/440298/26752599/236b7ed6-4886-11e7-806f-80d0ac675f77.png)

**Comment by bjoe2k4** [comment]
I agree, on such a painfully slow site, an additional connection probably doesn't matter. However, on a very fast site, standard dynamic injection or fetch-inject with connection re-usage is probably the way to go.
How about detecting external uri by testing for `//` and implementing some logic to add credentials based on that?

**Comment by jhabdas** [comment]
Let me put some thought into it and get back to you. Meanwhile if you'd like to open a new issue to help identify the opportunity I'd love to have a succinct thread in place where we can do some collaborative brainwork.

Aside: The only edge case I can think of with using // to determine external URLs is linked here (and it's an oddball case for certain,  but relevant because I plan to put fetch injection inside the theme): https://github.com/comfusion/after-dark/issues/37
